### PR TITLE
Add a conversation memory that combines a (optionally persistent) vectorstore history with a token buffer

### PR DESCRIPTION
**Pull request by lstein**
**langchain: ConversationVectorStoreTokenBufferMemory**

-**Description:** This PR adds ConversationVectorStoreTokenBufferMemory. It is similar in concept to ConversationSummaryBufferMemory. It maintains an in-memory buffer of messages up to a preset token limit. After the limit is hit timestamped messages are written into a vectorstore retriever rather than into a summary. The user's prompt is then used to retrieve relevant fragments of the previous conversation. By persisting the vectorstore, one can maintain memory from session to session.
-**Issue:** n/a
-**Dependencies:** none
-**Twitter handle:** Please no!!!
- [X] **Add tests and docs**:  I looked to see how the unit tests were written for the other ConversationMemory modules, but couldn't find anything other than a test for successful import. I need to know whether you are using pytest.mock or another fixture to simulate the LLM and vectorstore. In addition, I would like guidance on where to place the documentation. Should it be a notebook file in docs/docs?

- [X] **Lint and test**: I am seeing some linting errors from a couple of modules unrelated to this PR.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

File: libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+        split_chunk_size: (optional, 1000) Token chunk split size
+                          for long messages generated by the AI
+
+        previous_history_template: (optional) Template used to format
+                                   the contents of the prompt history
+
+
+    Example using ChromaDB:
+
```
Comment (isahers1) [inline libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py:66]: small nit, can we add the .. code-block:: python here?

Comment (lstein) [reply libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py:66]: This is done. I'm fairly new to docstrings. I hope I did it right?

**Comment by lstein** [comment]
BTW, I'm using this routinely with persistence in my personal chatbots and am constantly amazed at how frequently the chatbot recalls something relevant from a previous conversation. For small (7b) local models, this really works much better than the conversation summary chain.

File: libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py (hunk 1 above)
Comment (isahers1) [reply libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py:66]: So sorry for the small nit, you have to use "..code-block:: python". You can look at libs/community/langchain_community/chat_models/anthropic.py and see the doctoring for ChatAnthropic to see how its done

Comment (lstein) [reply libs/langchain/langchain/memory/token_buffer_vectorstore_memory.py:66]: Thank you for your patience! I hope I have it right now.

**Comment by lstein** [comment]
The `memory/test_imports.py` unit test was failing due to a circular import issue. I have fixed by changing the import order.

**Comment by lstein** [comment]
@isahers1 All tests are passing, documentation looks good. At what point does this get merged with the main branch — or is more needed to be done?

**Comment by isahers1** [comment]
Just merged, sorry for the delay!

**Pull request merged**